Investigate and fix php 7.3+ crash when copying a method using `new stdClass` (one of many php 7.3 crashes)

This would require looking into what changed in how constants are represented in PHP, with all platforms and settings (NTS vs ZTS, 32 vs 64 bit, etc.)

Things to look into:

- Read all of https://github.com/php/php-src/blob/master/UPGRADING.INTERNALS and see what applies. This may keep changing until php 7.3 is stable
- Are there more changes needed to handle specific opcode types?
- Are there more changes needed to handle the new constant/literals representation when copying methods?
- Does reference counting logic in runkit7 need to be changed?
- Should runkit be setting different flags or checking the flags (Related to pass_two() of zend_opcode.h)
- What's the simplest test case to reproduce this?

This is a follow up to #126 . This was noticed after rearranging the allocated literals into a single group.

```
Build complete.
Don't forget to run 'make test'.


=====================================================================
PHP         : /path/to/php-7.3.0alpha2-debug-ztsinstall/bin/php 
PHP_SAPI    : cli
PHP_VERSION : 7.3.0alpha2
ZEND_VERSION: 3.3.0-dev
PHP_OS      : Linux - Linux student3 4.15.0-24-generic #26~16.04.1-Ubuntu SMP Fri Jun 15 14:35:08 UTC 2018 x86_64
INI actual  : /path/to/runkit7/tmp-php.ini
More .INIs  :   
---------------------------------------------------------------------
PHP         : /path/to/php-7.3.0alpha2-debug-ztsinstall/bin/phpdbg 
PHP_SAPI    : phpdbg
PHP_VERSION : 7.3.0alpha2
ZEND_VERSION: 3.3.0-dev
PHP_OS      : Linux - Linux student3 4.15.0-24-generic #26~16.04.1-Ubuntu SMP Fri Jun 15 14:35:08 UTC 2018 x86_64
INI actual  : /path/to/runkit7/tmp-php.ini
More .INIs  : 
---------------------------------------------------------------------
CWD         : /path/to/runkit7
Extra dirs  : 
VALGRIND    : valgrind-3.11.0
=====================================================================
Running selected tests.
TEST 1/1 [tests/bug10053.phpt]
========OUT========
Runkit Method: 1
object(stdClass)#2 (1) {
  ["e"]=>
  int(4)
}
Runkit Method: 2
object(stdClass)#2 (1) {
  ["e"]=>
  int(9)
}

Notice: Undefined property: runkit_two::$ in /path/to/runkit7/tests/bug10053.php on line 7
Segmentation fault

Termsig=11
========DONE========

========MEM========
==6150== Invalid read of size 8
==6150==    at 0xCD2B04: zend_string_hash_val (zend_string.h:99)
==6150==    by 0xCD5584: zend_hash_find_bucket (zend_hash.c:559)
==6150==    by 0xCDAB03: zend_hash_find (zend_hash.c:2071)
==6150==    by 0xCA4374: zend_lookup_class_ex (zend_execute_API.c:885)
==6150==    by 0xCA593C: zend_fetch_class_by_name (zend_execute_API.c:1390)
==6150==    by 0xD45CC4: ZEND_NEW_SPEC_CONST_UNUSED_HANDLER (zend_vm_execute.h:8712)
==6150==    by 0xDAB622: execute_ex (zend_vm_execute.h:55368)
==6150==    by 0xDB00EE: zend_execute (zend_vm_execute.h:59962)
==6150==    by 0xCBF3DC: zend_execute_scripts (zend.c:1564)
==6150==    by 0xBF69A7: php_execute_script (main.c:2467)
==6150==    by 0xDB3081: do_cli (php_cli.c:1002)
==6150==    by 0xDB4513: main (php_cli.c:1395)
==6150==  Address 0x8 is not stack'd, malloc'd or (recently) free'd
==6150== 
==6150== 
==6150== Process terminating with default action of signal 11 (SIGSEGV)
==6150==  Access not within mapped region at address 0x8
==6150==    at 0xCD2B04: zend_string_hash_val (zend_string.h:99)
==6150==    by 0xCD5584: zend_hash_find_bucket (zend_hash.c:559)
==6150==    by 0xCDAB03: zend_hash_find (zend_hash.c:2071)
==6150==    by 0xCA4374: zend_lookup_class_ex (zend_execute_API.c:885)
==6150==    by 0xCA593C: zend_fetch_class_by_name (zend_execute_API.c:1390)
==6150==    by 0xD45CC4: ZEND_NEW_SPEC_CONST_UNUSED_HANDLER (zend_vm_execute.h:8712)
==6150==    by 0xDAB622: execute_ex (zend_vm_execute.h:55368)
==6150==    by 0xDB00EE: zend_execute (zend_vm_execute.h:59962)
==6150==    by 0xCBF3DC: zend_execute_scripts (zend.c:1564)
==6150==    by 0xBF69A7: php_execute_script (main.c:2467)
==6150==    by 0xDB3081: do_cli (php_cli.c:1002)
==6150==    by 0xDB4513: main (php_cli.c:1395)
==6150==  If you believe this happened as a result of a stack
==6150==  overflow in your program's main thread (unlikely but
==6150==  possible), you can try to increase the size of the
==6150==  main thread stack using the --main-stacksize= flag.
==6150==  The main thread stack size used in this run was 8388608.
========DONE========
LEAK&FAIL runkit_method_copy() function [tests/bug10053.phpt] 
=====================================================================
Number of tests :    1                 1
Tests skipped   :    0 (  0.0%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    0 (  0.0%) (  0.0%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests leaked    :    1 (100.0%) (100.0%)
Tests passed    :    0 (  0.0%) (  0.0%)
---------------------------------------------------------------------
Time taken      :    5 seconds
=====================================================================

=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
runkit_method_copy() function [tests/bug10053.phpt]
=====================================================================

=====================================================================
LEAKED TEST SUMMARY
---------------------------------------------------------------------
runkit_method_copy() function [tests/bug10053.phpt]
=====================================================================
Makefile:133: recipe for target 'test' failed
make: *** [test] Error 1
```

--- a/README.md
+++ b/README.md
@@ -2,7 +2,7 @@
 ============================================================================================
 
 For all those things you.... probably shouldn't have been doing anyway.... but surely do!
-__Now with partial support for PHP7.0, 7.1, and 7.2!__ (function/method manipulation is recommended only for unit testing. **Does not work with PHP 7.3 yet**.
+__Now with partial support for PHP7.0, 7.1, and 7.2!__ (function/method manipulation is recommended only for unit testing. (PHP 7.3 currently has minor bugs in `runkit_import` and internal function manipulation)
 
 [![Build Status](https://secure.travis-ci.org/runkit7/runkit7.png?branch=master)](http://travis-ci.org/runkit7/runkit7)
 [![Build Status (Windows)](https://ci.appveyor.com/api/projects/status/3jwsf76ge0yo8v74/branch/master?svg=true)](https://ci.appveyor.com/project/TysonAndre/runkit7/branch/master)
@@ -10,11 +10,6 @@ __Now with partial support for PHP7.0, 7.1, and 7.2!__ (function/method manipula
 [Building and installing runkit in unix](#building-and-installing-runkit7-in-unix)
 
 [Building and installing runkit in Windows](#building-and-installing-runkit7-in-windows)
-
-Current Build Status
---------------------
-
-In 7.0.x and 7.1.x and 7.2.0: 0 failing tests, 4 expected failures (constant manipulation in same file), 61 skipped tests(for disabled property and import), and 95 passing tests.
 
 Compatibility: PHP7.0 to PHP 7.2
 --------------------------------

--- a/package.xml
+++ b/package.xml
@@ -18,8 +18,8 @@ Define customized superglobal variables for general purpose use.
  </lead>
  <date>2018-07-04</date>
  <version>
-  <release>1.0.5b3</release>
-  <api>1.0.5b3</api>
+  <release>1.0.6</release>
+  <api>1.0.6</api>
  </version>
  <stability>
   <release>alpha</release>
@@ -27,8 +27,7 @@ Define customized superglobal variables for general purpose use.
  </stability>
  <license uri="http://www.opensource.org/licenses/BSD-3-Clause">BSD License (3 Clause)</license>
  <notes>
-- Fix incorrect reference counting of class names in parameter and return type signatures for `runkit_function_*()` and `runkit_method_*()` (https://github.com/runkit7/runkit7/issues/123).
-  This bug only affected php 7.2.0+
+- Fix crashes seen in php 7.3 alphas when function and method manipulation is used. (https://github.com/runkit7/runkit7/issues/126, https://github.com/runkit7/runkit7/issues/141)
  </notes>
  <contents>
   <dir name="/">
@@ -265,6 +264,21 @@ Define customized superglobal variables for general purpose use.
  <providesextension>runkit</providesextension>
  <extsrcrelease />
  <changelog>
+  <release>
+   <date>2018-07-04</date>
+   <version>
+    <release>1.0.6</release>
+    <api>1.0.6</api>
+   </version>
+   <stability>
+    <release>alpha</release>
+    <api>alpha</api>
+   </stability>
+   <license uri="http://www.opensource.org/licenses/BSD-3-Clause">BSD License (3 Clause)</license>
+   <notes>
+- Fix crashes seen in php 7.3 alphas when function and method manipulation is used. (https://github.com/runkit7/runkit7/issues/126, https://github.com/runkit7/runkit7/issues/141)
+   </notes>
+  </release>
   <release>
    <date>2018-07-04</date>
    <version>

--- a/runkit_functions.c
+++ b/runkit_functions.c
@@ -189,9 +189,8 @@ static inline void php_runkit_destroy_misplaced_internal_function(zend_function 
 /* {{{ php_runkit_set_opcode_constant
 		for absolute constant addresses, creates a local copy of that literal.
 		Modifies op's contents. */
-static void php_runkit_set_opcode_constant(const zval* literals, znode_op* op, zval* literalI) {
-	debug_printf("php_runkit_set_opcode_constant(%llx, %llx, %d), USE_ABS=%d", (long long)literals, (long long)literalI, (int)sizeof(zval), ZEND_USE_ABS_CONST_ADDR);
-	// TODO: ZEND_PASS_TWO_UPDATE_CONSTANT???
+static void php_runkit_set_opcode_constant(const zval* literals, znode_op* op, const zval* literalI) {
+	debug_printf("php_runkit_set_opcode_constant(%llx, %llx, %d), USE_ABS=%d\n", (long long)literals, (long long)literalI, (int)sizeof(zval), ZEND_USE_ABS_CONST_ADDR);
 #if ZEND_USE_ABS_CONST_ADDR
 	RT_CONSTANT_EX(literals, *op) = literalI;
 #else
@@ -205,12 +204,12 @@ static void php_runkit_set_opcode_constant(const zval* literals, znode_op* op, z
 /* {{{ php_runkit_set_opcode_constant_relative
 		for absolute constant addresses, creates a local copy of that literal.
 		Modifies op's contents. */
-static void php_runkit_set_opcode_constant_relative(const zend_op_array *op_array, const zend_op* opline, znode_op* op, zval* literalI) {
-	debug_printf("php_runkit_set_opcode_constant(%llx, %llx, %d), USE_ABS=%d", (long long)literals, (long long)literalI, (int)sizeof(zval), ZEND_USE_ABS_CONST_ADDR);
-	// TODO: ZEND_PASS_TWO_UPDATE_CONSTANT???
+static void php_runkit_set_opcode_constant_relative(const zend_op_array *op_array, const zend_op* opline, znode_op* op, const zval* literalI) {
+	debug_printf("php_runkit_set_opcode_constant_relative(%llx, %d), USE_ABS=%d\n", (long long)(long long)literalI, (int)sizeof(zval), ZEND_USE_ABS_CONST_ADDR);
 #if ZEND_USE_ABS_CONST_ADDR
 	RUNKIT_RT_CONSTANT(op_array, opline, *op) = literalI;
 #else
+	debug_printf("opcodes=%llx literals=%llx op=%llx literalI=%llx", (long long)op_array->opcodes, (long long) op_array->literals, (long long) op, (long long)literalI);
 	// TODO: Assert that this is in a meaningful range.
 	// TODO: is this ever necessary for relative constant addresses?
 	// Opposite of ((zval*)(((char*)(opline)) + (int32_t)(node).constant))
@@ -298,7 +297,9 @@ static zend_op *runkit_allocate_opcode_copy(const zend_op_array * const op_array
 #if PHP_VERSION_ID >= 70300
 	// Adjustment for https://github.com/runkit7/runkit7/issues/126
 	// I assume that PHP's pass_two() from Zend/zend_opcache.c would be done by now
-	if (!(ZEND_USE_ABS_CONST_ADDR) && op_array->literals) {
+	if (!(ZEND_USE_ABS_CONST_ADDR) &&
+			(op_array->fn_flags & ZEND_ACC_DONE_PASS_TWO) &&
+			op_array->literals) {
 		// Allocate enough space for op_array->opcode_copy and op_array->last_literal()
 		size_t bytes_to_allocate = ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_op) * op_array->last, 16) +
 			sizeof(zval) * op_array->last_literal;
@@ -317,9 +318,10 @@ static zval *runkit_allocate_literals(const zend_op_array * const op_array, zend
 #if PHP_VERSION_ID >= 70300
 	// Adjustment for https://github.com/runkit7/runkit7/issues/126
 	// I assume that PHP's pass_two() from Zend/zend_opcache.c would be done by now
-	if (!(ZEND_USE_ABS_CONST_ADDR)) {
+	if (!(ZEND_USE_ABS_CONST_ADDR) &&
+			(op_array->fn_flags & ZEND_ACC_DONE_PASS_TWO)) {
 		// Reuse the extra space allocated for runkit_allocate_opcode_copy
-		return (zval*)(((char*)op_array->opcodes) + ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_op) * op_array->last, 16));
+		return (zval*)(((char*)opcode_copy) + ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_op) * op_array->last, 16));
 	}
 #else
 	(void)opcode_copy;
@@ -450,8 +452,10 @@ static void php_runkit_function_copy_ctor_same_type(zend_function *fe, zend_stri
 
 		debug_printf("op_array.literals = %llx, last_literal = %d\n", (long long)op_array->literals, op_array->last_literal);
 		if (op_array->literals) {
+			debug_printf("   old op_array.literals = %llx, opcodes=%llx\n", (long long)op_array->literals, (long long)op_array->opcodes);
 			i = op_array->last_literal;
 			literals = runkit_allocate_literals(op_array, opcode_copy);
+			debug_printf("copied op_array.literals = %llx, opcodes=%llx\n", (long long)literals, (long long)opcode_copy);
 			while (i > 0) {
 				uint32_t k;
 
@@ -473,21 +477,28 @@ static void php_runkit_function_copy_ctor_same_type(zend_function *fe, zend_stri
 					if (opcode_copy[k].op2_type == IS_CONST) {
 						debug_printf("op2: %llx, %llx\n", (long long)RT_CONSTANT_EX(literals, opcode_copy[k].op2), (long long)&(op_array->literals[i]));
 					}*/
-					debug_printf("old constant = %d\n", new_op->op1.constant);
 					// TODO: This may be completely unnecessary on 64-bit systems. This may be broken on 32-bit systems.
 #ifdef RT_CONSTANT_EX
 					if (new_op->op1_type == IS_CONST && RT_CONSTANT_EX(literals, new_op->op1) == &op_array->literals[i]) {
+						debug_printf("old op1 constant = %d\n", new_op->op1.constant);
 						php_runkit_set_opcode_constant(literals, &(new_op->op1), &literals[i]);
+						debug_printf("new op1 constant = %d\n", new_op->op1.constant);
 					}
 					if (new_op->op2_type == IS_CONST && RT_CONSTANT_EX(literals, new_op->op2) == &op_array->literals[i]) {
+						debug_printf("old op2 constant = %d\n", new_op->op2.constant);
 						php_runkit_set_opcode_constant(literals, &(new_op->op2), &literals[i]);
+						debug_printf("new op2 constant = %d\n", new_op->op2.constant);
 					}
 #else
-					if (new_op->op1_type == IS_CONST && RUNKIT_RT_CONSTANT(op_array, new_op, new_op->op1) == &op_array->literals[i]) {
+					if (new_op->op1_type == IS_CONST && RUNKIT_RT_CONSTANT(op_array, &op_array->opcodes[i], new_op->op1) == &op_array->literals[i]) {
+						debug_printf("old op1 constant = %d\n", new_op->op1.constant);
 						php_runkit_set_opcode_constant_relative(op_array, new_op, &(new_op->op1), &literals[i]);
+						debug_printf("new op1 constant = %d\n", new_op->op1.constant);
 					}
-					if (new_op->op2_type == IS_CONST && RUNKIT_RT_CONSTANT(op_array, new_op, new_op->op2) == &op_array->literals[i]) {
+					if (new_op->op2_type == IS_CONST && RUNKIT_RT_CONSTANT(op_array, &op_array->opcodes[i], new_op->op2) == &op_array->literals[i]) {
+						debug_printf("old op2 constant = %d\n", new_op->op2.constant);
 						php_runkit_set_opcode_constant_relative(op_array, new_op, &(new_op->op2), &literals[i]);
+						debug_printf("new op2 constant = %d\n", new_op->op2.constant);
 					}
 #endif
 				}
@@ -552,7 +563,10 @@ static void php_runkit_function_copy_ctor_same_type(zend_function *fe, zend_stri
 		}
 	}
 
-	fe->common.fn_flags &= ~ZEND_ACC_DONE_PASS_TWO;
+	// I'm not sure why this was originally added in upstream -- Removing this isn't causing test failures in php 7.
+	// NOTE: If this line is uncommented, then PHP 7.3 would fail during cleanup of op arrays,
+	// because Zend's function destructor expects functions that didn't complete pass two to have a separate allocated array for literals.
+	// fe->common.fn_flags &= ~ZEND_ACC_DONE_PASS_TWO;
 	fe->common.prototype = fe;
 }
 /* }}} */

--- a/runkit_import.c
+++ b/runkit_import.c
@@ -613,6 +613,7 @@ PHP_FUNCTION(runkit_import)
 			zval *parent_name = RT_CONSTANT(new_op_array, new_op_array->opcodes[opline_num - 1].op2);
 			zval *key = parent_name + 1;
 			zend_class_entry *pce;
+			// TODO: Figure out why this assertion fails in php 7.3
 			ZEND_ASSERT(Z_TYPE_P(parent_name) == IS_STRING);
 
 			// TODO: Check if this is the same in php 7.0

